### PR TITLE
DRAFT: Preload contract on boot

### PR DESCRIPTION
**Pull request by jakmeier**
This is a draft implementation for  #7570. Submitted as draft PR for collaborative debugging.

cc @Longarithm 

Right now, this implementation gives me 259 threads from what I can see in the logs. In `iotop` I see at least 100 IO threads actively reading from disk with various speeds between 0 and 1000KB/s. 100KB/s looks like a rough median value.

With that I would expect that the full 1.1GB in the test account is fetched in roughly 2 minutes. But it still hasn't finished a single thread in a full hour.

The number of DB requests I have in the log surpassed 2 millions after about 1.5h but still it is not done. But that means even though it has read at least several GBs from disk, it still has only read 2 millions trie nodes. The RocksDB overhead is much more than I expect. Since we expect something like 10 million accounts, it looks like my current implementation simply does not scale enough. Or hopefully there is just something buggy in the implementation. But I am running out of ideas.

## Implementation Overview
On boot, right after starting the client actors, we send a new message of type `PreloadAccountData` to the client actor. 

Then create a `TrieIterator` and seek the root of the state subtrie of the account. From there, create a `HeavySubTrieIterator` that iterates over the selected subtrie in DFS but stops as soon as a leftover subtrie is lighter than 10MB as declared by its `memory_usage` field. When it reaches such a position, it returns a `TrieIterator` that will go over every element of the subtrie and continue with the next sibling.

So the output of `HeavySubTrieIterator` is a series of `TrieIterator`, each of them sufficiently small that we should be able to traverse it in a matter of seconds. The final step is to spawn a new thread for each `TrieIterator` and traverse it.

Note that we simply read each value once and then disgard it. Each thread has its own `Trie` instance but they share the `Arc<Mutex<TrieCacheInner>>`. This will load everything into the same shard cache as a side-effect.

Also, I am using plain std threads. I also tried using tokio threads but it introduces new problems and solves none. There is nothing that could be made async, as RocksDB does not support it. But we can consider moving it tokio threads if we want, once it works as intended.

Threads spawning generally could be improved, among other things. But it is not worth fleshing out if the performance is not what we need.

### CODE REVIEW

**Comment by jakmeier** [comment]
Hm, I think my archival node I used for testing is just in a bad state. I noticed that it is compacting all the while, which already takes 25% of the IO capacity. And we have seen recently that RocksDB performance can be reduced significantly in such cases. I am now fully compacting the state of my node and will try again afterwards.

File: chain/client/src/client_actor.rs [hunk 1]
```diff
@@ -1961,6 +1961,89 @@ impl Handler<StateSplitResponse> for ClientActor {
...
+            let n = TrieIterator { trie: &trie, trail, key_nibbles }.count();
+            trace!(target: "store", "Preload subtrie at {root} done");
+            n
+        });
+        handles.push(handle);
+    }
+
+    let mut n = 0;
+    for handle in handles {
+        n += handle.join().expect("thread failed");
```
Comment (matklad) [inline chain/client/src/client_actor.rs:2040]: `std::thread::scope` is a thing now! And, yeah, given that this is IO bound, I don't tihnk we should re-use our existing rayon's CPU pool.

**Comment by jakmeier** [comment]
I had compaction running for 12h+ but it didn't finish. But I saw GBs being moved in the SST files.

Now I tested with a fresh RocksDB from an archival snapshot. Good news is now I see no compaction work being done in the background, so I would hope now it works to preload the data quickly. But it still takes too long. It has been running for an hour but didn't finish. If this should be an option at all, it must finish in a few minutes.

I could try doing it on an RPC node setup but I'm not sure it matters too much. If it doesn't work for archival nodes, then I tend towards not doing it at all.

**Comment by Longarithm** [comment]
@jakmeier what if we reduce the subtrie size to 1 MB, etc? Do we know the number of nodes in each subtrie?

**Comment by matklad** [comment]
Drive by comment, but something does not add up for me here. If I understand the work here correctly, what we are doing is loading a couple of gigs from disk into RAM. It doesn't make sense that it takes sooo long. 

One think that occured to me is that, as we are using hases as key, we essentially hammer the database in random reads. We could perhaps traverse the trie not in a trie order, but in the order of node hashes? That is, we can keep in-memory frontier of hashes to visit, and make sure we traverse it in lexicographic order?

**Comment by Longarithm** [comment]
> That is, we can keep in-memory frontier of hashes to visit, and make sure we traverse it in lexicographic order?

Interesting idea. I doubt that it can improve the speed significantly, but I don't see a way to prove or disprove this without experiment.
Looks like a complex mathematical task - "Assume that we keep a sorted set of, say, 100K hashes, and we always extract first 1K hashes and add new 1K children hashes to the set. How many hashes read in 1K batch fall in the same rocksdb block?"

In the long term, I would consider
* experiment with "fast-kv-store" to see what the overhead will be
* add prefixes to rocksdb representation of trie keys which indicate "freshness" of the key. We thought about prefixing key with block height. I'm actually curious if rocksdb supports "alternative sorting" for entries - so we wouldn't need to modify the key, but instead store block height in metadata by which keys will be sorted.

For RPC nodes it may even make sense to read the whole `DBCol::State` sequentially. But then we still don't know to which heights the read nodes belong to, and for archival nodes it doesn't work at all.

**Comment by jakmeier** [comment]
@matklad 
You are spot on, we are hammering random accesses due to hashes. And RocksDB moves blocks of 4KB in and out of the cache each time, with uncompression overhead etc.

The idea with iterating with a frontier and accessing the sorted values first is very interesting. Looking at the number involved here, I cannot see it work for archival nodes. We are selecting 1GB from several TBs with "perfect spread" thanks to cryptographic hash functions. I don't expect a lot of locality on disk. Even a factor of 2 improvement would surprise me, but we need more than that.

@akhi3030 
- The way I had it coded then, the remaining CPU threads are spinning . It's only 100% utilisation in `htop`. I expect them to be mostly idle  from an engineering point of view. Although I haven't looked at flamecharts, maybe there really is also a CPU bottleneck.
- I'm using gcloud "SSD persistent disk" which is what our terraform templates use for archival nodes. `fio` gives me read throughput of 803MiB/s. And the same write throughput.  I'm not quite sure why `iotop` showed me 100% IO utilisation, maybe I'm misinterpreting its output. I start to think maybe the 8 CPU cores are the limit here doing the RocksDB internal work. Which wouldn't make things much better tbh. Maybe Edvard's path could give us 30% but still it's not enough.

@Longarithm 
As far as my tests have shown so far, the plain KV store runs into the same problem of random accesses as soon as it hits a certain size. Enforcing data locality seems much more promising to me.

Enforcing data locality is actually something I wanted to discuss in the storage meeting for several weeks. It is similar to the "freshness" idea you mentioned. My idea is to somehow guarantee data locality within each account's state. This could be done within RocksDB by prefixing all state column keys with the account id (or account id hash). Or even use a separate KV store per account instead of RocksDB. But the more I think about the second option, the more I think we will just end up re-implementing RocksDB.

**Comment by nagisa** [comment]
> I'm not quite sure why iotop showed me 100% IO utilisation, maybe I'm misinterpreting its output.

To the best of my knowledge, `iotop` measures I/O time, not throughput or anything else. If the block device is busy servicing requests for 1 second for the past 1 second, you will see 100% utilization during that time slice. However, this does not actually mean that the maximum throughput is achieved – many block devices, including hard disk drives (nowadays some of them include multiple read heads!), can handle concurrent requests and service them quicker than if it was servicing these same requests in a strict sequence. In either case `iotop` would show 100% utilization, though.

This sequencing will also have an increasingly disproportional impact as the latency between the host and disk increases.

Now, you _are_ utilizing multiple threads and intuitively that ought to result in reasonable I/O queue depths, but it is worth checking anyway – maybe there’s some locking, or an inappropriate I/O scheduler is being used, or…

**Comment by jakmeier** [comment]
this is no longer needed

**Comment by jakmeier** [comment]
this one might have been slow due to the problem we reverted in #7648 
in other words, it seems I was using RocksDB without SSE instructions and this likely caused this to be CPU bottlenecked much quicker than it would be when trying again on master today.
==> Do not take performance numbers from this experiment as golden truth

**Pull request closed without merging**